pkg/frontend/querymiddleware/stats.go

```diff
@@ -62,5 +65,41 @@ func (s queryStatsMiddleware) Do(ctx context.Context, req Request) (Response, er
 		}
 	}
 
+	if details := QueryDetailsFromContext(ctx); details != nil {
+		details.Start = time.UnixMilli(req.GetStart())
+		details.End = time.UnixMilli(req.GetEnd())
+		details.Step = time.Duration(req.GetStep()) * time.Millisecond
```

Comment:
I think there's a fondamental flaw here. According to the PR description, start/end are read to log the actual query time range and figure out whether a query only hits ingesters or store-gateway. However, the actual query time range is not just driven by start/end parameter but also by vector selector duration (e.g. think about an instant query where start=end=now but has `metric[30d]`).

I think a better approach may be storing the **min and max timestamp** in the querier stats, and then use such information to log the actual query timerange (which is a very interesting data point we're completely missing today).

Reply:
> I think a better approach may be storing the min and max timestamp in the querier stats, and then use such information to log the actual query timerange (which is a very interesting data point we're completely missing today).

i agree. Looking at the code which currently validates this I think we have to expose a method of the promQL engine https://github.com/grafana/mimir/blob/423e6f00a9f116e2fe5e0e318ac4fabd7cc5dcf3/vendor/github.com/prometheus/prometheus/promql/engine.go#L820

I will propose this upstream first.

In the meantime is it ok to merge the current approximation? It will be a few minutes off, but it will have the correct ballpark value

Reply:
I opened https://github.com/prometheus/prometheus/pull/13085 to export `findMinMaxTime`

Reply:
that got merged and into mimir quicker than I expected. I'm not integrating it in this PR